File: csharp/Streaming.cs
using System.Net.Http.Headers;
using System.Runtime.CompilerServices;
using System.Text;
using System.Text.Json;

namespace MisarMail;

/// <summary>
/// One decoded Server-Sent Events frame.
///
/// MisarMail emits unnamed frames — <c>data: {...}</c> with no <c>event:</c>
/// line — so <see cref="Event"/> is normally null. <see cref="Data"/> is the
/// decoded JSON, or null when the payload was not JSON; <see cref="Raw"/> is
/// always the payload exactly as received.
/// </summary>
public sealed record MisarMailStreamEvent(string? Event, JsonElement? Data, string Raw);

public sealed partial class MisarMailClient
{
    /// <summary>
    /// <c>POST /api/ai/generate-email/stream</c> — token-by-token generation.
    /// </summary>
    /// <example>
    /// <code>
    /// await foreach (var e in client.Streaming_GenerateEmailAsync(new { prompt = "…" }))
    ///     Console.Write(e.Data?.GetProperty("delta").GetString());
    /// </code>
    /// </example>
    public IAsyncEnumerable<MisarMailStreamEvent> Streaming_GenerateEmailAsync(
        object options, CancellationToken ct = default) =>
        SseStreamAsync(HttpMethod.Post, "/ai/generate-email/stream", options, ct);

    /// <summary>
    /// <c>GET /api/campaigns/{id}/send-stream</c> — live send progress.
    /// </summary>
    public IAsyncEnumerable<MisarMailStreamEvent> Streaming_CampaignSendAsync(
        string campaignId, CancellationToken ct = default) =>
        SseStreamAsync(
            HttpMethod.Get,
            $"/campaigns/{Uri.EscapeDataString(campaignId)}/send-stream",
            null,
            ct);

    /// <summary>
    /// Opens an SSE connection and yields each frame as it arrives.
    ///
    /// Uses the API base, since both streaming routes sit outside <c>/v1</c>.
    /// Frames end at a blank line and may span several <c>data:</c> lines; the
    /// <c>[DONE]</c> sentinel ends the stream and is not yielded.
    ///
    /// Deliberately not retried: replaying a stream that failed mid-flight
    /// would duplicate whatever the caller already consumed.
    /// </summary>
    private async IAsyncEnumerable<MisarMailStreamEvent> SseStreamAsync(
        HttpMethod method,
        string path,
        object? body,
        [EnumeratorCancellation] CancellationToken ct = default)
    {
        using var request = new HttpRequestMessage(method, _apiBase + path);
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _apiKey);
        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("text/event-stream"));
        if (body is not null)
            request.Content = new StringContent(
                JsonSerializer.Serialize(body), Encoding.UTF8, "application/json");

        HttpResponseMessage response;
        try
        {
            // ResponseHeadersRead is what makes this a stream rather than a
            // buffered read of the whole body.
            response = await _httpClient
                .SendAsync(request, HttpCompletionOption.ResponseHeadersRead, ct)
                .ConfigureAwait(false);
        }
        catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException)
        {
            throw new MisarMailNetworkException(ex.Message);
        }

        using (response)
        {
            int status = (int)response.StatusCode;

            if (!response.IsSuccessStatusCode)
            {
                string raw = await response.Content.ReadAsStringAsync(ct).ConfigureAwait(false);
                if (IsPlanLimit(raw))
                    throw BuildPlanLimitError(status, raw, response);
                throw new MisarMailException(status, ExtractError(raw));
            }

            await using var stream = await response.Content
                .ReadAsStreamAsync(ct).ConfigureAwait(false);
            using var reader = new StreamReader(stream, Encoding.UTF8);

            string? eventName = null;
            var dataLines = new List<string>();

            while (true)
            {
                string? line;
                try
                {
                    line = await reader.ReadLineAsync(ct).ConfigureAwait(false);
                }
                catch (Exception ex) when (ex is IOException or HttpRequestException)
                {
                    throw new MisarMailNetworkException($"stream read: {ex.Message}");
                }
                if (line is null) break;

                if (line.Length == 0)
                {
                    var frame = BuildFrame(ref eventName, dataLines);
                    if (frame is null) continue;              // keepalive or empty
                    if (ReferenceEquals(frame, DoneFrame)) yield break;
                    yield return frame;
                    continue;
                }
                // Comments are keepalives.
                if (line[0] == ':') continue;

                if (line.StartsWith("event:", StringComparison.Ordinal))
                    eventName = line[6..].Trim();
                else if (line.StartsWith("data:", StringComparison.Ordinal))
                    dataLines.Add(line[5..].StartsWith(' ') ? line[6..] : line[5..]);
            }

            // A final frame with no trailing blank line.
            var tail = BuildFrame(ref eventName, dataLines);
            if (tail is not null && !ReferenceEquals(tail, DoneFrame))
                yield return tail;
        }
    }

    /// <summary>Sentinel identity for <c>data: [DONE]</c>, compared by reference.</summary>
    private static readonly MisarMailStreamEvent DoneFrame = new(null, null, "[DONE]");

    /// <summary>
    /// Consumes the accumulated lines into one frame, clearing them either way.
    /// Returns <see cref="DoneFrame"/> for the sentinel and null for an empty frame.
    /// </summary>
    private static MisarMailStreamEvent? BuildFrame(ref string? eventName, List<string> dataLines)
    {
        if (dataLines.Count == 0)
        {
            eventName = null;
            return null;
        }
        string raw = string.Join("\n", dataLines);
        string? name = eventName;
        dataLines.Clear();
        eventName = null;

        if (raw == "[DONE]") return DoneFrame;

        JsonElement? data = null;
        try
        {
            using var doc = JsonDocument.Parse(raw);
            data = doc.RootElement.Clone();
        }
        catch (JsonException) { /* not JSON — Raw still carries it */ }

        return new MisarMailStreamEvent(name, data, raw);
    }
}

File: rust/src/streaming.rs
//! Server-Sent Events support.
//!
//! MisarMail exposes two streams, both **outside** the `/v1` prefix, so they use
//! the API base rather than the versioned one:
//!
//! - `POST /api/ai/generate-email/stream`
//! - `GET  /api/campaigns/{id}/send-stream`
//!
//! Both emit *unnamed* frames — `data: {...}` with no `event:` line — and
//! terminate with the sentinel `data: [DONE]`, which is consumed here rather
//! than surfaced. MisarReach uses named events instead, so the two products
//! deliberately do not share a parser.

use futures_util::{Stream, StreamExt};
use serde_json::Value;

use crate::MisarMailError;

/// One decoded SSE frame.
#[derive(Debug, Clone, PartialEq)]
pub struct StreamEvent {
    /// The SSE `event:` name, when the server sends one. Normally `None`.
    pub event: Option<String>,
    /// Decoded JSON payload, or `Value::String` when the payload was not JSON.
    pub data: Value,
}

/// Index of the blank line ending the first complete frame, or `None`.
pub(crate) fn frame_end(buffer: &str) -> Option<usize> {
    match (buffer.find("\n\n"), buffer.find("\r\n\r\n")) {
        (None, None) => None,
        (Some(lf), None) => Some(lf),
        (None, Some(crlf)) => Some(crlf),
        (Some(lf), Some(crlf)) => Some(lf.min(crlf)),
    }
}

/// Outcome of parsing one frame.
pub(crate) enum Frame {
    /// A payload to hand to the caller.
    Event(StreamEvent),
    /// The `[DONE]` sentinel — stop reading.
    Done,
    /// A keepalive comment or an empty frame.
    Skip,
}

pub(crate) fn parse_frame(frame: &str) -> Frame {
    let mut event = None;
    let mut data_lines: Vec<&str> = Vec::new();

    let normalised = frame.replace("\r\n", "\n");
    for line in normalised.lines() {
        if line.is_empty() || line.starts_with(':') {
            continue;
        }
        if let Some(rest) = line.strip_prefix("event:") {
            event = Some(rest.trim().to_owned());
        } else if let Some(rest) = line.strip_prefix("data:") {
            data_lines.push(rest.strip_prefix(' ').unwrap_or(rest));
        }
    }
    if data_lines.is_empty() {
        return Frame::Skip;
    }
    let raw = data_lines.join("\n");
    if raw == "[DONE]" {
        return Frame::Done;
    }
    let data = serde_json::from_str(&raw).unwrap_or(Value::String(raw));
    Frame::Event(StreamEvent { event, data })
}

/// Consumes a streaming response and yields decoded frames, stopping at `[DONE]`.
///
/// Deliberately not retried: replaying a stream that failed mid-flight would
/// duplicate whatever the caller already consumed.
pub type EventStream =
    std::pin::Pin<Box<dyn Stream<Item = Result<StreamEvent, MisarMailError>> + Send>>;

pub(crate) fn frames_from(resp: reqwest::Response) -> EventStream {
    let bytes = resp.bytes_stream();
    Box::pin(futures_util::stream::unfold(
        (Box::pin(bytes), String::new(), false),
        |(mut bytes, mut buffer, done)| async move {
            if done {
                return None;
            }
            loop {
                // Serve anything already buffered before reading more.
                if let Some(end) = frame_end(&buffer) {
                    let frame: String = buffer.drain(..end).collect();
                    // Drop the blank-line separator that follows.
                    while buffer.starts_with('\n') || buffer.starts_with('\r') {
                        buffer.remove(0);
                    }
                    match parse_frame(&frame) {
                        Frame::Event(e) => return Some((Ok(e), (bytes, buffer, false))),
                        Frame::Done => return None,
                        Frame::Skip => continue,
                    }
                }

                match bytes.next().await {
                    Some(Ok(chunk)) => buffer.push_str(&String::from_utf8_lossy(&chunk)),
                    Some(Err(e)) => {
                        return Some((Err(MisarMailError::Network(e)), (bytes, buffer, true)))
                    }
                    None => {
                        // Socket closed: emit a trailing frame if one is complete.
                        if buffer.trim().is_empty() {
                            return None;
                        }
                        let frame = std::mem::take(&mut buffer);
                        return match parse_frame(&frame) {
                            Frame::Event(e) => Some((Ok(e), (bytes, String::new(), true))),
                            _ => None,
                        };
                    }
                }
            }
        },
    ))
}
